review_routes/v3/results_routes.py

```diff
@@ -809,3 +933,152 @@ async def get_alignment_scores(
+    baseline_ids : list, optional
+        A list of baseline revision ids to compare against.
+    book : str, optional
+        Restrict results to one book.
+    chapter : int, optional
+        Restrict results to one chapter. If set, book must also be set.
+    verse : int, optional
+        Restrict results to one verse. If set, book and chapter must also be set.
+    page : int, optional
+        The page of results to return. If set, page_size must also be set.
+    page_size : int, optional
+        The number of results to return per page. If set, page must also be set.
+    """
+
+    await validate_parameters(book, chapter, verse)
+
+    # Remove baseline ids for revisions belonging to the same version as the revision or reference
+    revision_version_subquery = db.query(BibleRevision.bible_version_id).filter(BibleRevision.id == revision_id).subquery()
+    reference_version_subquery = db.query(BibleRevision.bible_version_id).filter(BibleRevision.id == reference_id).subquery()
+    revisions_with_same_version = db.query(BibleRevision.id).filter(
+        BibleRevision.bible_version_id.in_([revision_version_subquery.as_scalar(), reference_version_subquery.as_scalar()])
+    ).all()
+    ids_with_same_version = [result.id for result in revisions_with_same_version]
+    baseline_ids = [id for id in baseline_ids if id not in ids_with_same_version]
+
+    # Initialize base query
+    (
+        main_assessment_query,
+        total_count,
+        assessment_id,
+    ) = await build_missing_words_main_query(
+        revision_id,
+        reference_id,
+        book,
+        chapter,
+        verse,
+        page,
+        page_size,
+        db,
+    )
+
+    main_assessment_results = db.execute(main_assessment_query).fetchall()
+    (
+        baseline_assessment_query,
+        assessment_to_baseline_id,
+    ) = await build_missing_words_baseline_query(
+        reference_id,
+        baseline_ids,
+        book,
+        chapter,
+        verse,
+        db,
+    )
+    baseline_assessment_results = db.execute(baseline_assessment_query).fetchall()
+    df_main = pd.DataFrame(main_assessment_results)
+    if baseline_assessment_results:
+        df_baseline = pd.DataFrame(baseline_assessment_results).drop(columns=["id"])
+    else:
+        df_baseline = pd.DataFrame(
+            columns=[
+                "book",
+                "chapter",
+                "verse",
+                "source",
+                "target",
+            ]
+        )
+    df_baseline.loc[:, "baseline_id"] = df_baseline["assessment_id"].apply(
+        lambda x: assessment_to_baseline_id[x]
+    )
+    df_baseline = df_baseline.drop(columns=["assessment_id"])
+    joined_df = pd.merge(
+        df_main, df_baseline, on=["book", "chapter", "verse", "source"], how="left"
+    )
+
+    df = (
+        joined_df.groupby(["book", "chapter", "verse", "source"])
+        .apply(
+            lambda x: pd.Series(
```

Comment:
i think there are 3 lambda function pandas conversions in the flow, if the performance of this route is not great, we can look into vectorizing those operations

Reply:
Great, yeah that would be good. Ideally I would have everything computed within SQL, but when I joined the two main queries, it wasn't working - the SQLAlchemy didn't seem to be quite creating the right SQL for the join. So I ended up doing two SQL queries and joining them in Pandas, which isn't ideal.

So yeah, at some point it might be good to optimize this.